src/noInapplicableLifeCycleRule.ts

```diff
@@ -0,0 +1,159 @@
+    'OnDestroy',
+  ]);
+
+  static CMP_OR_DIR_ONLY: Set<string> = difference(Rule.CMP_OR_DIR, Rule.SERVICE);
+
+  public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+    return this.applyWithWalker(new LifeCycleDecoratorWalker(sourceFile, this.getOptions()));
+  }
+
+}
+
+export class LifeCycleDecoratorWalker extends Ng2Walker {
+
+  public inheritanceForest = new InheritanceForest();
+
+  public possibleFailures = new Map<string, Lint.RuleFailure>();
+
+  private addPossibleFailure(node: ts.ClassDeclaration, disallowedHookNames: string[]): void {
+    const hooks = disallowedHookNames
+      .map(hook => ({
+        name: hook,
+        path: this.inheritanceForest.getPath(node.name.text, hook),
+      }))
+      .map(hook => {
+        if (hook.path.length == 2) {
+          hook.path = null;
+        }
+        return hook;
+      });
+    const failure = this.createFailure(
+      node.name.getStart(),
+      node.name.getWidth(),
+      sprintf.apply(this, [Rule.FAILURE_NOT_CMP_OR_DIR_FACTORY(hooks), node.name.text]));
+    this.possibleFailures.set(node.name.text, failure);
+  }
+
+  public visitClassDeclaration(node: ts.ClassDeclaration) {
+    const className: string = node.name.text;
+
+    const interfaces = getHeritage(node, SyntaxKind.current().ImplementsKeyword);
+    const interfaceNames = interfaces.map(getHeritageNames);
+    const baseClassNames = getHeritage(node, SyntaxKind.current().ExtendsKeyword).map(getHeritageNames);
+
+    this.inheritanceForest.add(className, [...baseClassNames, ...interfaceNames]);
+    const heritageNames = this.inheritanceForest.getHeritage(className);
+    // console.log(className, '-->', heritageNames);
+
+    const decoratorNames = getDecoratorNames(node);
+
+    // Add a new possible failure
+    if (decoratorNames.every(name => name != 'Component' && name != 'Directive')) {
+      // not a component, not a directive
+
+      // if true:  not a service (we disallow OnDestroy as well)
+      // if false: a service (we allow OnDestroy but still look for others)
+      const checkAgainst = decoratorNames.every(name => name != 'Injectable') ?
+        Rule.CMP_OR_DIR :
+        Rule.CMP_OR_DIR_ONLY;
+
+      const disallowedHooks = Array.from(intersection(new Set(heritageNames), checkAgainst));
+      if (disallowedHooks.length > 0) {
+        // it implements a life cycle which can be only applied to component or directive
+        this.addPossibleFailure(node, disallowedHooks);
+      }
+    }
+
+    // Remove an earlier added possible failure
+    heritageNames.forEach(base => {
+      if (this.possibleFailures.has(base)) {
+        // We marked this as a possible failure
+        // Now we know it's not because it's extended (so the error is "propagated"  down the
+        // inheritance tree until the leaves where it's either resolved or stays there
+        this.possibleFailures.delete(base);
+      }
+    });
+    this['failures'] = []; // Clear failures?
```

Comment:
So I couln't find any hook or a way to collect some data and after walking is complete to do something with it. This method is called by `applyWalker` and it just returns the failures. I can see how to get the failures, but not how to change them after I've added them.